discovery-provider/src/solana/solana_program_indexer.py

```diff
@@ -0,0 +1,263 @@
+        @param session: DB session
+        @param tx_sig: transaction signature to look up
+        """
+        raise Exception("Must be implemented in subclass")
+
+    @abstractmethod
+    def get_latest_slot(self):
+        """
+        Return the highest slot for this indexer's associated entity -
+        for example, track listens checks the Play table while the user
+        bank indexer checks user_bank_txs
+        """
+        raise Exception(BASE_ERROR)
+
+    @abstractmethod
+    async def parse_tx(self, tx_sig):
+        """
+        Parse an individual transaction, this will vary based on the program being indexed
+        @param tx_sig: transaction signature to be parsed
+        """
+        tx_info = self._solana_client_manager.get_sol_tx_info(tx_sig)
+        result: TransactionInfoResult = tx_info["result"]
+        return {"tx_sig": tx_sig, "tx_metadata": {}, "result": result}
+
+    @abstractmethod
+    async def fetch_ipfs_metadata(self, parsed_transactions):
+        """
+        Fetch all metadata objects in parallel (if required). Certain indexing tasks will not require this step and can skip appropriately
+        @param parsed_transactions: Array of transactions containing deserialized information, ideally pointing to a metadata object
+        """
+        return {}
+
+    async def process_txs_batch(self, tx_sig_batch_records):
+        self.msg(f"Parsing {tx_sig_batch_records}")
+        futures = []
+        tx_sig_futures_map: Dict[str, Dict] = {}
+        for tx_sig in tx_sig_batch_records:
+            future = asyncio.ensure_future(self.parse_tx(tx_sig))
+            futures.append(future)
+        for future in asyncio.as_completed(futures, timeout=100000):
+            try:
+                future_result = await future
+                self.msg(f"{future_result}")
+                tx_sig_futures_map[future_result["tx_sig"]] = future_result
+            except asyncio.CancelledError:
+                # Swallow cancelled requests
+                pass
+
+        # Committing to DB
+        parsed_transactions = []
+        for tx_sig in tx_sig_batch_records:
+            parsed_transactions.append(tx_sig_futures_map[tx_sig])
+
+        # Fetch metadata in parallel
+        metadata_dictionary = self.fetch_ipfs_metadata(parsed_transactions)
+
+        self.validate_and_save_parsed_tx_records(
+            parsed_transactions, metadata_dictionary
+        )
+
+    @abstractmethod
+    def validate_and_save_parsed_tx_records(
+        self, parsed_transactions, metadata_dictionary
+    ):
+        """
+        Based parsed transaction information, generate and save appropriate database changes. This will vary based on the program being indexed
+        @param parsed_transactions: Array of transaction signatures in order
+        @param metadata_dictionary: Dictionary of remote metadata
+        """
+        raise Exception(BASE_ERROR)
+
+    def get_transactions_to_process(self):
```

Comment:
```suggestion
    def get_tx_batches_to_process(self):
```
based on description - i think this is more suggestive of array of arrays. wdyt?

Reply:
great call